Add search to all ubuntu.com docs

carkod opened the pull request:
## Done

Search for all ubuntu.com docs

## QA

Theses Checks:
- Search something that yields many results e.g. application and click the next and previous pagination links
- Search something that don't yield results e.g. asdf and check links are correct


In the following URLs:

https://ubuntu-com-9375.demos.haus/ceph/docs
https://ubuntu-com-9375.demos.haus/server/docs
https://ubuntu-com-9375.demos.haus/core/docs
https://ubuntu-com-9375.demos.haus/openstack/docs


## Issue / Card

Fixes #9049 #9050 #9051 #9381

cassiocassio commented:
![Screenshot 2021-03-12 at 09.41.16.png](https://images.zenhubusercontent.com/59391bc151aa2c305780dbd6/b20daa1a-199f-4308-91ef-55c84e6c981e)

- "Search on ubuntu.com" text crunches into icon

removing "on" would do it.

(Is the change from "Seach" - on the live site deliberate?)

nottrobin commented:
Both "search on Ubuntu.com" and "search Ubuntu.com" are also inaccurate.

It is in fact a search across most of canonical's public websites.

anasereijo commented:
Hello, I added a design needed label as I think this needs a design perspective. 

- I think the search input shouldn’t have a light grey background. When the page has secondary navigation (also in light grey) makes it look like part of the same thing. I suggest having a white background instead.

<img width="1121" alt="Screenshot 2021-03-12 at 10 19 06" src="https://user-images.githubusercontent.com/36884067/110926519-67479700-831c-11eb-9769-3fdecdf4ec27.png">


- I am also not sure about the size of the search input. It's very wide IMO, but you probably replicated what we have on [Vanilla](https://vanillaframework.io/docs/examples/layouts/documentation), right?

- This is more a question for UX, but wondering if it would be useful to have filters? Something like we have on [tutorials](https://ubuntu.com/tutorials?q=install). Probably not necessary but just wondering... @elioqoshi let us know your thoughts.

- search on ubuntu.com is now broken on mobile:

how it looks like on your demo:
<img width="479" alt="Screenshot 2021-03-12 at 10 17 19" src="https://user-images.githubusercontent.com/36884067/110926387-3c5d4300-831c-11eb-9865-91a29c4a0105.png">


how it looks like currently on ubuntu.com:
<img width="477" alt="Screenshot 2021-03-12 at 10 17 12" src="https://user-images.githubusercontent.com/36884067/110926413-43845100-831c-11eb-93d1-0102e9f27392.png">

wgx commented:
Search placeholder text of "Search our sites" might resolve the issue @nottrobin mentioned.

![image](https://user-images.githubusercontent.com/930099/110931687-76c9de80-8322-11eb-9a40-46b81cd79d74.png)

wgx commented:
Overall, I think the usability of:

- "Master" search in the global nav to search everything, and
- "Sub section" search in the hero area with clear indication of _what areas_ the user is searching within..

..is fine and seems clear. 👍🏽

anthonydillon commented:
If it helps at all, next iteration I am planning to implement the collapsed/expending search in the header so that might help the UX.

cassiocassio commented:
+1 to @nottrobin 's wording, since e.g a search for "multipass"  gives results for  https://multipass.run/docs 

I propose "Search Ceph Docs" would be better in other field - the exact path is an implementation detail

carkod commented:
> +1 to @nottrobin 's wording, since e.g a search for "multipass" gives results for https://multipass.run/docs
> 
> I propose "Search Ceph Docs" would be better in other field - the exact path is an implementation detail

Thanks. Done.

carkod commented:
> Search placeholder text of "Search our sites" might resolve the issue @nottrobin mentioned.
> 
> ![image](https://user-images.githubusercontent.com/930099/110931687-76c9de80-8322-11eb-9a40-46b81cd79d74.png)

Done. Thanks

carkod commented:
> Hello, I added a design needed label as I think this needs a design perspective.
> 
> * I think the search input shouldn’t have a light grey background. When the page has secondary navigation (also in light grey) makes it look like part of the same thing. I suggest having a white background 

- [x] Done 

> 
> <img alt="Screenshot 2021-03-12 at 10 19 06" width="1121" src="https://user-images.githubusercontent.com/36884067/110926519-67479700-831c-11eb-9769-3fdecdf4ec27.png">
> 
> * This is more a question for UX, but wondering if it would be useful to have filters? Something like we have on [tutorials](https://ubuntu.com/tutorials?q=install). Probably not necessary but just wondering... @elioqoshi let us know your thoughts.
> * search on ubuntu.com is now broken on mobile:
> 
- Not possible at this point. We don't have any data to classify the content (no categories, or tags or any kind of taxonomy), we only have [navigation data](https://discourse.ubuntu.com/t/about-the-documentation-category/17250).

> how it looks like on your demo:
> <img alt="Screenshot 2021-03-12 at 10 17 19" width="479" src="https://user-images.githubusercontent.com/36884067/110926387-3c5d4300-831c-11eb-9865-91a29c4a0105.png">
> 
> how it looks like currently on ubuntu.com:
> <img alt="Screenshot 2021-03-12 at 10 17 12" width="477" src="https://user-images.githubusercontent.com/36884067/110926413-43845100-831c-11eb-93d1-0102e9f27392.png">

Fixed.

Regarding the search width, [juju](https://juju.is/docs) is the one I used as this was the main docs people requested features from, but yes, currently Vanilla does not provide a way to customize the search width, I could do something very specific for this one, like set it to a `max-width: 40em` which seems to be what we have for paragraphs, but I believe this could be a relevant discussion for Vanilla.

anasereijo commented:
@carkod looks great! Thanks 👍

nottrobin commented on webapp/app.py:
```diff
@@ -436,9 +436,21 @@ def takeovers_index():
         url_prefix=url_prefix,
     ),
-    document_template="/templates/docs/discourse.html",
+    document_template="/templates/docs/server/docs.html",
```
Shouldn't these be at e.g. `/server/docs.html`?

nottrobin commented on webapp/app.py:
```diff
@@ -436,9 +436,21 @@ def takeovers_index():
+    build_search_view(
+        session=session,
+        site="ubuntu.com/server/docs",
+        template_path="templates/docs/server/search.html",
```
Similarly:

```suggestion
        template_path="server/docs/search.html",
```

nottrobin commented on templates/templates/docs/openstack/docs.html:
```diff
@@ -0,0 +1,149 @@
+        sideNavigation.classList.add('is-expanded');
+      } else {
+        sideNavigation.classList.remove('is-expanded');
+        sideNavigation.classList.add('is-collapsed');
+      }
+    }
+  }
+
+  /**
+    Attaches event listeners for the side navigation toggles
+    @param {HTMLElement} sideNavigation The side navigation element.
+  */
+  function setupSideNavigation(sideNavigation) {
+    var toggles = [].slice.call(sideNavigation.querySelectorAll('.js-drawer-toggle'));
+
+    toggles.forEach(function (toggle) {
+      toggle.addEventListener('click', function (event) {
+        event.preventDefault();
+        var sideNav = document.getElementById(toggle.getAttribute('aria-controls'));
+
+        if (sideNav) {
+          toggleDrawer(sideNav, !sideNav.classList.contains('is-expanded'));
+        }
+      });
+    });
+  }
+
+  /**
+    Attaches event listeners for all the side navigations in the document.
+    @param {String} sideNavigationSelector The CSS selector matching side navigation elements.
+  */
+  function setupSideNavigations(sideNavigationSelector) {
+    // Setup all side navigations on the page.
+    var sideNavigations = [].slice.call(document.querySelectorAll(sideNavigationSelector));
+
+    sideNavigations.forEach(setupSideNavigation);
+  }
+
+  setupSideNavigations('.p-side-navigation, [class*="p-side-navigation--"]');
+</script>
```
You've only used a shared include for the one tiny piece, of the `_search-box.html`, but basically the whole of this template is identical to the other 3 isn't it? You've used an include for the only piece that's different!

I would have thought that instead of creating `templates/docs/shared/_search-box.html`, you want to create `templates/shared/docs/_document.html` and put all of this content in there, including the search box.

```suggestion
  {% with 
    search_action="/openstack/docs/search",
    siteSearch="https://ubuntu.com/openstack/docs",
    placeholder="Search on Charmed Openstack Docs"
  %}
    {% include "shared/docs/_document.html" %}
  {% endwith %}
```

nottrobin commented on templates/templates/docs/openstack/search.html:
```diff
@@ -0,0 +1,85 @@
+
+{% block outer_content %}
+
+  <div class="p-strip is-shallow">
```
Same here, this template is almost identical to the other search results pages for the other docs, so shouldn't it be in an include like `shared/docs/_search-results.html`?

nottrobin reviewed (changes requested):
Thanks Carlos. I've made some suggestions about how you could restructure your templates to save some duplication.

nottrobin commented on templates/templates/search.html:
```diff
@@ -1,5 +1,5 @@
 <form action="/search" id="google-appliance-search-form" class="header-search">
 		<label for="edit-keys" class="accessibility-aid">Search</label>
-    <input type="search" maxlength="255" name="q" id="edit-keys" class="form-text" placeholder="Search" value="{{ query }}" />
+    <input type="search" maxlength="255" name="q" id="edit-keys" class="form-text" placeholder="Search on ubuntu.com" value="{{ query }}" />
```
This still says "Search on ubuntu.com"

```suggestion
    <input type="search" maxlength="255" name="q" id="edit-keys" class="form-text" placeholder="Search our sites" value="{{ query }}" />
```

nottrobin reviewed (approved):
Just one "Search on ubuntu.com" that still need changing. Other than that, this looks good. Much neater, thanks. Once you've changed this you can merge.

(I'm not crazy about the way our template folders are structured. `templates/templates` is nuts. I hate that we have both `templates/shared` and `templates/docs/shared`. But I don't think you should do anything about this in this PR - you're just following existing patterns. I'll try to find time to do a bit of a template restructure PR myself instead.)

nottrobin commented:
Oh, https://ubuntu-com-9375.demos.haus/openstack/docs gives me a 500 error...

nottrobin commented:
When QAing, I came across this:

![image](https://user-images.githubusercontent.com/519935/111336038-c84cd300-866c-11eb-978d-a32048580403.png)

I think, given that this search results page has its own template, we can do better with this message. Maybe 'Sorry we couldn't find "fish" in server docs' all in one line. Or something.

carkod commented:
> When QAing, I came across this:
> 
> ![image](https://user-images.githubusercontent.com/519935/111336038-c84cd300-866c-11eb-978d-a32048580403.png)
> 
> I think, given that this search results page has its own template, we can do better with this message. Maybe 'Sorry we couldn't find "fish" in server docs' all in one line. Or something.

Done

> When QAing, I came across this:
> 
> ![image](https://user-images.githubusercontent.com/519935/111336038-c84cd300-866c-11eb-978d-a32048580403.png)
> 
> I think, given that this search results page has its own template, we can do better with this message. Maybe 'Sorry we couldn't find "fish" in server docs' all in one line. Or something.

Done. Also fixed an issue with searching empty query.

nottrobin reviewed (approved):
Thanks for all the hard work @carkod. This looks good to me now.
